### ISSUE TITLE
Languages should be a Mapping

### ISSUE THREAD

**Issue by jayvdb** [coala/coala#5658]
#5657 first. Then make Languages a Mapping and Iterable class (see collections)

**Comment by corona10** [comment]
So it should be like this below?

``` python
l = Languages(['C#', 'Python 2.7'])
isinstance(l, abc.Mapping)
>>> True
```

**Comment by jayvdb** [comment]
> l = Languages(['C#', 'Python 2.7'])

yes.  The silly part about the 'tuple' based approach is that the constructor accepts *fuzzy* inputs, but then it produces a collection where it is not possible to correlate the inputs with the contents of the collection.

Consider

> l = Languages(['C#', 'Python 2.7', 'CS'])

Should produce the same collection as your example?  CS = C# ?
If yes, then how does the caller know where CS is in the tuple?
If the caller doesnt know where the CS is in the tuple, the `Languages` object has very limited usefulness.

Or does it contain the CS language twice?

Or should it produce an error?

Or should it produce a mapping, so that all of the inputs are mapped to a `Language` object.

Or is there some other option I havent considered?

I believe a mapping is the most logical approach.  The following would then be True

> assert l['CS'] == l['C#']

Ideally also

> assert l['CS'] is l['C#']

An open question is whether this collection should be immutable.  When it was a tuple, it was immutable.  I have no opinion on that yet. ;-)

**Comment by corona10** [comment]
@jayvdb 
Okay, this problem as same as a search engine problem.
If someone queries `cpp` or `c++`, a search engine should provide the same information.
Now I understand why you want to provide Mapping interface for `Languages`.

My current PR doesn't look like prper way to solve the issue you want,
I will update my PR with better one :)
